rotkehlchen/externalapis/alchemy.py

```diff
@@ -0,0 +1,422 @@
+        for entry in result:
+            if (
+                ('symbol' in asset_identifier and entry['symbol'] != asset_identifier['symbol']) or
+                (
+                    'address' in asset_identifier and
+                    entry['network'] != asset_identifier['network'] and
+                    entry['address'] != asset_identifier['address']
+                )
+            ):
+                continue
+
+            if 'error' in entry:
+                log.warning(
+                    f'Queried Alchemy Price API from {from_asset.identifier} to {to_asset.identifier}. '  # noqa: E501
+                    f'API returned error: {entry["error"]}',
+                )
+                return ZERO_PRICE
+
+            if not (prices := entry['prices']):
+                log.warning(
+                    f'Queried Alchemy Price API from {from_asset.identifier} to {to_asset.identifier}, '  # noqa: E501
+                    f'but received no prices in response',
+                )
+                return ZERO_PRICE
+
+            for price in prices:
+                if price['currency'] != 'usd':
+                    continue
+
+                try:
+                    return deserialize_price(price['value'])
+                except (DeserializationError, KeyError) as e:
+                    error_msg = f'missing key {e!s}' if isinstance(e, KeyError) else str(e)
+                    log.warning(
+                        f'Queried Alchemy Price API from {from_asset.identifier} to {to_asset.identifier}. '  # noqa: E501
+                        f'Failed to deserialize price data due to: {error_msg}',
+                    )
+                    return ZERO_PRICE
+
+        log.warning(
+            f'Queried Alchemy Price API from {from_asset.identifier} to {to_asset.identifier}, '
+            f'but no USD price entry was found in the response',
+        )
+        return ZERO_PRICE
+
+    @staticmethod
+    def _deserialize_historical_price(
+            result: dict[str, Any],
+            timestamp: Timestamp,
+            asset_identifier: dict[str, str],
+            from_asset: Asset,
+            to_asset: Asset,
+    ) -> Price:
+        """Deserializes historical price data from Alchemy API response.
+
+        If price data is missing, invalid, or there's an error, ZERO_PRICE is returned.
+        Otherwise, the deserialized price is returned.
+        """
+        if (
+            ('symbol' in asset_identifier and result['symbol'] != asset_identifier['symbol']) or
+            (
+                'address' in asset_identifier and
+                result['network'] != asset_identifier['network'] and
+                result['address'] != asset_identifier['address']
+            )
+        ):
+            log.warning(
+                f'Queried Alchemy API from {from_asset.identifier} to {to_asset.identifier}, '
+                f'but received mismatched symbol in response',
+            )
+            return ZERO_PRICE
```

Comment:
this logic seems very similar to the one in the other function. Can't this be abstracted somehow?